File: Java/StrFSeconds.java
// import java.util.stream.Collectors;
// import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

import java.math.BigDecimal;
import java.math.RoundingMode;

//import java.util.*;


class StrFSeconds {
    public static void main(String[] args) {
        // See div mod calculations, using a double or float loses
        // precision? 4000 seconds returns 4000000010.099 milliseconds.
        //
        // >>> System.out.println(4000/.000001);
        // 4000000010.099029064178466796875
        //
        // Maybe proceed to BigDecimal for more precision:
        //
        // >>> BigDecimal bD = new BigDecimal(4000/.000001);
        // >>> System.out.println(bD.toPlainString());
        // 4000000000
        //
        // BigDecimal Floor
        //BigDecimal bd1 = new BigDecimal(7200);
        //BigDecimal bd2 = new BigDecimal(3600);
        //BigDecimal bd2 = new BigDecimal(3600.0000000000000001);
        //System.out.println( bd1.divide(bd2, 30, RoundingMode.CEILING));
        //System.out.println( bd1.divideToIntegralValue(bd2));

        // unitSize = unitSize.add( secondsDivMod.divide(unit.getValue(), 2, RoundingMode.FLOOR));

        // Why is Java 'String' type written in capital letter while
        // 'int'"' is not?
        // https://stackoverflow.com/a/4006311
        //

        long startTime = System.currentTimeMillis();


        double seconds;
        String formatStringResult;

        //seconds=90;
        //seconds=1;
        //seconds=4000;
        seconds=4000.1234567890;
        //seconds=4000;

        formatStringResult = StrFSeconds(seconds, "%h:%m:%s", 10);
        System.out.println(formatStringResult);

        formatStringResult = StrFSeconds(seconds, "%f");
        System.out.println(formatStringResult);


        long elapsedTime = System.currentTimeMillis() - startTime;
        System.out.println("[Elapsed in " + elapsedTime + "ms]");
    }


    public static String StrFSeconds(double seconds, String formatString, int nDecimal) {

        String smallestUnitInFormatString = null;
        BigDecimal unitSize = new BigDecimal(0);
        String unitSizePlainString;
        BigDecimal secondsDivMod = new BigDecimal(seconds);

        //System.out.println(seconds);

        // Arrays in Java
        // Instead of a normal Hashmap create a LinkedHashMap that will
        // iterate in the order in which the entries were put into the map
        // https://stackoverflow.com/a/17910409
        // https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/LinkedHashMap.html

        LinkedHashMap<String,BigDecimal> units = new LinkedHashMap<>();
        units.put("w", new BigDecimal(604800));
        units.put("d", new BigDecimal(86400));
        units.put("h", new BigDecimal(3600));
        units.put("m", new BigDecimal(60));
        units.put("s", new BigDecimal(1));
        units.put("l", new BigDecimal(.001));
        units.put("f", new BigDecimal(.000001));

        // Determine the smallest time unit in the formatString, if no
        // smallest time unit found return formatString.
        for (Map.Entry<String,BigDecimal> unit : units.entrySet())
            if (formatString.indexOf("%"+unit.getKey()) != -1)
                smallestUnitInFormatString = unit.getKey();

        if (smallestUnitInFormatString == null)
            return formatString;

        // For every time unit in units...
        for (Map.Entry<String,BigDecimal> unit : units.entrySet()) {
            // Check if this time unit is in formatString, if not
            // continue.
            if (formatString.indexOf("%"+unit.getKey()) == -1)
                continue;

            // Since BigDecimal implementation, (re)set value of
            // unitSize.
            // https://coderanch.com/t/534944/java/setting-BigDecimal-variable
            unitSize = BigDecimal.ZERO;

            // divmod > div
            //
            // Get the quotient of this units seconds and the
            // available seconds.
            //
            // unit_size, seconds = divmod(seconds, unit['secs'])
            //
            // # Cast to int Math.floor to prevent 'error: incompatible types:
            // possible lossy conversion from double to double'
            // # Do not cast to int! Causes unexpected results on large
            // numbers. (error: integer number too large)
            //unitSize = (int)Math.floor(seconds / unit.getValue());
            //unitSize = Math.floor(seconds / unit.getValue());
            //unitSize = unitSize.add(new BigDecimal(Math.floor(seconds / unit.getValue())));

            // Divide with precision set to 0
            unitSize = unitSize.add(secondsDivMod.divide(unit.getValue(), 0, RoundingMode.FLOOR));


            // divmod > mod
            //
            // Assign modulus (remainder) to seconds
            //
            //seconds = seconds % unit.getValue();
            secondsDivMod = secondsDivMod.remainder(unit.getValue());

            //System.out.println("Time unit: %" + unit.getKey());
            //System.out.println(seconds);
            //System.out.println(unit.getValue());
            //System.out.println(seconds / unit.getValue());
            //System.out.println(new BigDecimal(Math.floor(seconds / unit.getValue())));
            //System.out.println(unitSize);
            //System.out.println();

            // Smallest unit, add the remaing seconds as fractions of
            // unit size.
            if (unit.getKey() == smallestUnitInFormatString)
                // unit_size += 1 / unit['secs'] * seconds
                // unitSize += 1 / unit.getValue() * seconds;
                // unitSize = unitSize.add(new BigDecimal(1 / unit.getValue() * seconds));
                unitSize = unitSize.add(secondsDivMod.divide(unit.getValue(), nDecimal, RoundingMode.FLOOR));


            // calculations are done, proceed with formatting

            // Disable scientific notation from double
            // https://stackoverflow.com/a/47809639
            // BigDecimal d = new BigDecimal(unitSize);
            // unitSizeString = d.toPlainString();
            // unitSizeString = String.valueOf(unitSize);
            //unitSizePlainString = new BigDecimal(unitSize).toPlainString();
            unitSizePlainString = unitSize.toPlainString();

            //System.out.println(unitSize);

            formatString = formatString.replace("%"+unit.getKey(), unitSizePlainString);
        }


        return formatString;
    }

    // Default parameter values?
    // Use method overloading.
    // https://stackoverflow.com/a/1038401
    public static String StrFSeconds( double seconds, String formatString) {
        return StrFSeconds( seconds, formatString, 3);
    }

    public static String StrFSeconds( double seconds) {
        return StrFSeconds( seconds, "%h2:%m2:%s", 3);
    }

}

File: PHP/strfseconds.php
<?php
// strfseconds.php
//




// echo strfseconds( 550.194812, '%o %s.%f', 0) . "\n";

// echo strfseconds( 550.194812, '%f') . "\n";



//echo (strfseconds(3600, '%o %h2:%m2:%s2') . "\n");

// echo (strfseconds((float)550.194812, '%o %h:%m:%s.%f')) . "\n";
//echo (strfseconds((float)550.194810, '%f')) . "\n";

//echo (strfseconds( 123.75, '%s6')) . "\n";

// echo (strfseconds((float)550.194812, '%f')) . "\n";

// echo (strfseconds((float)550.194813123456, '%f6')) . "\n";

//echo (strfseconds((float)550.1948132, '%f')) . "\n";
//echo (strfseconds((float)550.1948331234, '%f')) . "\n";
//echo (strfseconds((float)123456789, '%s')) . "\n";
// echo (strfseconds((float)10.194812, '%o %h:%m:%s.%f')) . "\n";

//echo fmod( 550.194812, .0000001). "\n";


function strfseconds( float $seconds, string $formatstring = '%d2:%m2:%s2', int $ndecimal = 3) {

    $smallest_unit_in_formatstring = null;

	if ($seconds < 0)
		throw new Exception(
			"Seconds must be greater than or equal to 0");
	if ($ndecimal < 0)
		throw new Exception(
			"Decimal must be greater than or equal to 0");

    $units = array(
        'w' => (float)604800, // week
        'd' => (float)86400, // day
        'h' => (float)3600, // hour
        'm' => (float)60, // minute
        's' => (float)1, // second
        'f' => (float).000001 // microsecond
    );

    // Replace %o by the value passed for seconds
    $formatstring = str_replace( '%o', $seconds, $formatstring);

    // Determine the smallest time unit in the formatstring and
    // set variable accordingly. If no smallest time unit found
    // return unaltered formatstring.
    foreach ( $units as $unit_name => $unit_secs)
        if ( strpos($formatstring, '%' . $unit_name) !== false)
            $smallest_unit_in_formatstring = $unit_name;
    if ( $smallest_unit_in_formatstring == null)
        return $formatstring;


    // For every time unit ...
    foreach ( $units as $unit_name => $unit_secs) {

        // Check if the time unit is defined in the formatstring, if
        // not continue to the next unit.
        if ( strpos($formatstring, '%' . $unit_name) === false)
            continue;

        // Execute divmod on seconds for this units size: divide the
        // number of seconds by this unit's number of seconds.
        // The unit_size is assigned the quotient, seconds is assigned
        // the remainder.
        // unit_size, seconds = divmod(seconds, unit['secs'])
        $unit_size = floor( $seconds / $unit_secs );

        // On very small numers (microseconds) fmod leaves fractions
        // that should not be there. Rewrote to manual calculation.
        // $seconds = fmod( $seconds, $unit_secs );
        //
        // Float limited precision may show one ms off:
        //
        // >>> echo strfseconds( 550.194812, '%o %s.%f', 0);
        // 550.194812 550.194811
        //
        // >>> echo 550.194812 - 550;
        // 0.19481199999996

        $seconds = $seconds - ( $unit_size * $unit_secs );

        // If this unit is the smallest unit add the remaing seconds as
        // fractions of this time unit's size.
        if ( $smallest_unit_in_formatstring == $unit_name )
            $unit_size += 1 / $unit_secs * $seconds;


        // Calculations are done. Convert unit_size to string and format
        // appearance.


        $unit_size = (string)$unit_size;


        if ( $smallest_unit_in_formatstring == $unit_name ) {
            // Truncate decimals after ndecimal
            // https://stackoverflow.com/a/12710283
            // (string functions are available on float values)
            if ( ($posdecimal = strpos( $unit_size, '.')) !== false)
                $unit_size = floatval( substr( $unit_size, 0, $posdecimal + $ndecimal + 1));

            // Apply sprintf to ensure fixed number of decimals
            // (if ndecimal=3 '.5' should display as '.500')
            //
            $unit_size = sprintf( '%.' . $ndecimal . 'f', $unit_size);
        }

        // Left pad unit_size with zeroes
        $unit_rpad_length = '';
        if ( preg_match('/%' . $unit_name . '([0-9])/', $formatstring, $m) ) {
            // Assign the zeroes specifier parameter to $unit_rpad_length
            // (used also in final formatstring replacement)
            $unit_rpad_length = $m[1];
            // Check for decimals; decimals have difference length
            // value in str_pad().
            if ( strpos( $unit_size, '.' ) ) {
                $unit_size = str_pad(
                    $unit_size, strlen( $unit_size ) - strpos( $unit_size, '.' ) + $unit_rpad_length, '0', STR_PAD_LEFT);
            } else {
                $unit_size = str_pad(
                    $unit_size, $unit_rpad_length, '0', STR_PAD_LEFT);
            }
        }

        // Replace $unit_name . $unit_rpad_length by $unit_size.
        // Eg; replace '%h2' by '03'
        $formatstring = str_replace('%' . $unit_name . $unit_rpad_length, $unit_size, $formatstring);

    }

	return $formatstring;

}
